hw/ip/otbn/dv/otbnsim/sim/stats.py

```diff
@@ -0,0 +1,302 @@
+                if sym_addr <= address and sym_addr > func_addr:
+                    func_addr = sym_addr
+            symbol_name = self._addr_symbol_map[func_addr] + f"+{address - func_addr:#x}"
+
+        file_line = None
+        if self._elf_file.has_dwarf_info():
+            dwarf_info = self._elf_file.get_dwarf_info()
+            file_line = _dwarf_decode_file_line(dwarf_info, address)
+
+        add_info = []
+        if symbol_name:
+            add_info.append(f"{symbol_name}")
+        if file_line:
+            add_info.append(f"at {file_line[0]}:{file_line[1]}")
+
+        str = f"{address:#x}"
+        if add_info:
+            str += ' (' + ' '.join(add_info) + ')'
+        return str
+
+    def dump(self) -> str:
+        out = ""
+        out += "\n"
+        out += "Execution time\n"
+        out += "--------------\n"
+        out += self._dump_execution_time()
+        out += "\n\n"
+        out += "Instruction frequencies\n"
+        out += "-----------------------\n"
+        out += self._dump_instruction_histo()
+        out += "\n\n"
+        out += "Function call statistics\n"
+        out += "------------------------\n"
+        out += self._dump_function_call_stats()
+        out += "\n\n"
+        out += "Loop statistics\n"
+        out += "---------------\n"
+        out += self._dump_loop_stats()
+        out += "\n"
+
+        return out
+
+    def _dump_execution_time(self) -> str:
+        insn_count = self._stats.insn_count
+        stall_count = self._stats.stall_count
+        stall_percent = stall_count / insn_count * 100
+        cycles = insn_count + stall_count
+        time_ms = cycles / (self.FREQ_MHZ * 1e6) * 1e3
+
+        out = f"OTBN executed {insn_count} instructions in {cycles} cycles.\n"
+
+        out += f"The execution stalled for {stall_count} cycles "
+        out += f"({stall_percent:.01f} percent).\n"
+
+        out += f"The execution would take {time_ms:.02f} ms "
+        out += f"@ {self.FREQ_MHZ} MHz.\n"
+        return out
+
+    def _dump_instruction_histo(self) -> str:
+        return tabulate(self._stats.instruction_histo.most_common(),
+                        headers=['instruction', 'count']) + "\n"
+
+    def _dump_function_call_stats(self) -> str:
+        """ Dump function call statistics """
+        out = ""
+
+        # TODO: Add the start address as function?
+
+        # Build function call graphs and a call site index
+        # caller-indexed == forward, callee-indexed == reverse
+        # The call graphs are on function granularity; the call sites dictionary
+        # is indexed by the called function, but uses the call site as value.
+        callgraph = {}
+        rev_callgraph = {}
+        rev_callsites = {}
+        for c in self._stats.func_calls:
+            if c['caller_func'] not in callgraph:
+                callgraph[c['caller_func']] = Counter()
+            callgraph[c['caller_func']][c['callee_func']] += 1
```

Comment:
If the `Counter()` constructor isn't too expensive, this might be simpler as:
```python3
callgraph.setdefault(c['caller_func'], Counter())[c['callee_func']] += 1
```
?

Reply:
I find that less readable, the nested indexing is already quite dense.